pkg/didcomm/protocol/introduce/service.go

```diff
@@ -221,65 +155,74 @@ func (s *Service) startInternalListener() {
 			if err := s.InvitationReceived(event); err != nil {
 				logger.Errorf("listener invitation received: %s", err)
 			}
-		case <-s.stop:
-			s.wg.Done()
-
-			return
 		}
 	}
 }
 
 func logInternalError(err error) {
-	if _, ok := err.(*customError); !ok {
+	if _, ok := err.(customError); !ok {
 		logger.Errorf("go to abandoning: %v", err)
 	}
 }
 
-func (s *Service) doHandle(msg service.DIDCommMsg, outbound bool) (*metaData, error) {
+func contextID(msg service.DIDCommMsg) string {
+	contextID := msg.Metadata()[metaContextID]
+	if cID, ok := contextID.(string); ok && cID != "" {
+		return cID
+	}
+
+	return threadID(msg)
+}
+
+func threadID(msg service.DIDCommMsg) string {
+	if pthID := msg.ParentThreadID(); pthID != "" {
+		return pthID
+	}
+
 	thID, err := msg.ThreadID()
+	if errors.Is(err, service.ErrThreadIDNotFound) {
+		msg.(service.DIDCommMsgMap)["@id"] = uuid.New().String()
+		return msg.(service.DIDCommMsgMap)["@id"].(string)
+	}
+
 	if err != nil {
-		return nil, err
+		panic(err)
```

Comment:
Are you saying this is an impossible case?

Reply:
It can happen only when the message is invalid. The invalid message is a message with missing id. But it will be filtered by the messenger on the lower level. So, it will never happen. But, I also think that the `ThreadID()` function should not return an error. I would like to rewrite it. Instead of an error, an empty value will be returned. It will reduce this error handling and improve readability.